WordPress/src/main/java/org/wordpress/android/ui/notifications/NotificationsDetailActivity.java

```diff
@@ -581,31 +589,37 @@ public void onEventMainThread(NotificationEvents.NotificationsRefreshError error
         if (note == null) {
             // no note found
             showErrorToastAndFinish();
-            return;
         }
     }
 
     @Override
     public void onPositiveClicked(@NonNull String instanceTag) {
-        Fragment fragment = mAdapter.getItem(mViewPager.getCurrentItem());
-        if (fragment instanceof BasicFragmentDialog.BasicDialogPositiveClickInterface) {
-            ((BasicDialogPositiveClickInterface) fragment).onPositiveClicked(instanceTag);
+        if (mBinding != null && mAdapter != null) {
+            Fragment fragment = mAdapter.getItem(mBinding.viewpager.getCurrentItem());
+            if (fragment instanceof BasicFragmentDialog.BasicDialogPositiveClickInterface) {
+                ((BasicDialogPositiveClickInterface) fragment).onPositiveClicked(instanceTag);
+            }
         }
     }
 
     @Override
     public void onScrollableViewInitialized(int containerId) {
-        AppBarLayoutExtensionsKt.setLiftOnScrollTargetViewIdAndRequestLayout(mAppBarLayout, containerId);
+        if (mBinding != null) {
+            AppBarLayoutExtensionsKt.setLiftOnScrollTargetViewIdAndRequestLayout(mBinding.appbarMain, containerId);
+        }
     }
 
+    @SuppressWarnings("deprecation")
     private class NotificationDetailFragmentAdapter extends FragmentStatePagerAdapter {
         private final ArrayList<Note> mNoteList;
 
+        @SuppressWarnings("unchecked")
         NotificationDetailFragmentAdapter(FragmentManager fm, ArrayList<Note> notes) {
             super(fm, BEHAVIOR_RESUME_ONLY_CURRENT_FRAGMENT);
             mNoteList = (ArrayList<Note>) notes.clone();
         }
 
+        @NonNull
```

Comment:
I understand that the super method `getItem` is annotated as `@NonNull`, but the return value here is nullable, since `getDetailFragmentForNote` is nullable. I'm not sure why the compiler doesn't complain, but if the `Note` passed in as an argument is `null`, this will return `null`. Maybe we can assert that the note is not null? I.e. something like:

```diff
diff --git a/WordPress/src/main/java/org/wordpress/android/ui/notifications/NotificationsDetailActivity.java b/WordPress/src/main/java/org/wordpress/android/ui/notifications/NotificationsDetailActivity.java
index 1fcb956270..6bf0414f9e 100644
--- a/WordPress/src/main/java/org/wordpress/android/ui/notifications/NotificationsDetailActivity.java
+++ b/WordPress/src/main/java/org/wordpress/android/ui/notifications/NotificationsDetailActivity.java
@@ -68,6 +68,7 @@ import org.wordpress.android.widgets.WPViewPagerTransformer;
 import java.util.ArrayList;
 import java.util.HashMap;
 import java.util.Map;
+import java.util.Objects;
 
 import javax.inject.Inject;
 
@@ -622,7 +623,8 @@ public class NotificationsDetailActivity extends LocaleAwareActivity implements
         @NonNull
         @Override
         public Fragment getItem(int position) {
-            return getDetailFragmentForNote(mNoteList.get(position));
+            @NonNull Note note = Objects.requireNonNull(mNoteList.get(position), "Note cannot be null");
+            return getDetailFragmentForNote(note);
         }
 
         @Override
```

Wdyt?

Reply:
> I understand that the super method getItem is annotated as @NonNull, but the return value here is nullable, since getDetailFragmentForNote is nullable.

Hmmm, interesting find @mkevins ! 💯 

> I'm not sure why the compiler doesn't complain, but if the Note passed in as an argument is null, this will return null.

The compiler doesn't complain because it doesn't know whether `getDetailFragmentForNote(...)` is `@Nullable` or `@NonNull`. So, if you annotate that method as `@Nullable` it will not start to complain. AFAIA, this is expected. 👍 

> Maybe we can assert that the note is not null? I.e. something like:

Thank you for suggesting this change! 🥇 

Thinking about it, maybe it is better to declare that this method accept a `@NonNull Note note` and returns a `@NonNull Fragment` instead, then delete the whole `if (note == null) { return null; }` "unnecessary" guarding code. The result will be the same, doing that or having `requireNonNull(...)` we will get a NPE either way, but this other way it makes it explicit that we don't want `note` to be nullable code-wise, instead of having a comment dictating it, which we know, devs might miss. Wdyt? 🤔

Reply:
Btw, I want to note here as an FYI, that this work here is related to adding missing nullability annotations to view related `fields` first (via the `findViewById` to `ViewBinding` conversion), and then maybe any other left-over fields missing such nullability annotation.

As such, the scope of this PR wasn't to deal with adding missing nullability annotations on method parameters or return values. Thus, you didn't see me deal with this issue you reported in this comment. Also, AFAIA, we might have more such issues.

Going forward with these `findViewById` to `ViewBinding` conversion PRs, I would like to continue that path, only focusing on the `fields` side of the problem and ignore adding any missing nullability annotations on the `method` side. I want to do that so as to not explode the scope of a PR and achieve merging it to `trunk` faster.

Do you agree with this approach @mkevins ? 🤔

Reply:
> maybe it is better to declare that this method accept a @NonNull Note note and returns a @NonNull Fragment instead, then delete the whole if (note == null) { return null; } "unnecessary" guarding code.

Since this method is private, and it's only usage is here, I think this is a good approach. 👍 

>  instead of having a comment dictating it, which we know, devs might miss. Wdyt?

I'm not sure what you refer to by "comment", but I assume you mean the _message_ as the second argument to `Objects.requireNonNull`. The main advantage is that this helps pinpoint the NPE, and the reason. I'm not concerned about developers "missing" it, since it's intended for better stack traces (hard to miss there 😄). Even with the above suggested changes, we still need to assert that `mNoteList.get(position)` is `@NonNull`, so I think having the message is useful there. This, imo, is the _whole point_ of fail-fast. We want to fail as early (and as clearly) as possible when we are breaking contract.

Reply:
👋 @mkevins !

> Since this method is private, and it's only usage is here, I think this is a good approach. 👍

👍 

> I'm not sure what you refer to by "comment", but I assume you mean the message as the second argument to Objects.requireNonNull. The main advantage is that this helps pinpoint the NPE, and the reason. I'm not concerned about developers "missing" it, since it's intended for better stack traces (hard to miss there 😄).

Ah, for some reason I read the above as:

```java
@NonNull Note note = Objects.requireNonNull(mNoteList.get(position)); // Note cannot be null
```

🤦 😊 🤦 

> Even with the above suggested changes, we still need to assert that mNoteList.get(position) is @NonNull, so I think having the message is useful there. This, imo, is the whole point of fail-fast. We want to fail as early (and as clearly) as possible when we are breaking contract.

Oh yes, you are right about that and I was thinking about it too, was trying to figure out if `mNoteList.get(position)` will be null or not, that is, given the fact (or assumption) that a `Note` added to that list should never be null. But, I wanted to defer that investigation to another time.

Thus, I would still avoid adding the extra `Objects.requireNonNull(mNoteList.get(position), "Note cannot be null")` check there, because that would mean that this is now becoming the rule here, and we should strive doing that on every such instance of `Fragment getItem(int position) { ... }` in our project. But, I am not convinced yet that it is actually needed, I need to investigate if there are cases where `getItem(...)` will be called on the adapter but the list to get that item from might be empty, or else not having that item on that exact position, or else if that item can be null (so far my assumption is no to all of those questions). Does that make sense? 🤔

Also, this is another reason why I didn't want to go that deep in trying to solve every nullable problem while dealing with this `findViewById` to `ViewBinding` replacement, and only wanting to deal with `fields` related to that. And definitely not trying to Kotlinize any of those classes, as this will then force me to enter this rabbit hole anyway. Otherwise, as you can understand, this will get quickly out of hand, definitely out of scope. 🤷

Reply:
FYI: This is now done: 69601155de97c4bcce6fbe4dbba18cb3beea73d3

Reply:
> I would still avoid adding the extra Objects.requireNonNull(mNoteList.get(position), "Note cannot be null") check there, because that would mean that this is now becoming the rule here, and we should strive doing that on every such instance of Fragment getItem(int position) { ... } in our project.

I don't agree that adding this assertion means that it is "becoming a rule", and that doing so requires adding a similar check everywhere that something _appears_ similar throughout the codebase. This kind of rationale would prevent us from making small changes toward a larger goal. I think we should focus on this instance, and avoid prematurely generalizing.

>  But, I am not convinced yet that it is actually needed, I need to investigate if there are cases where getItem(...) will be called on the adapter but the list to get that item from might be empty, or else not having that item on that exact position, or else if that item can be null (so far my assumption is no to all of those questions). Does that make sense?

I don't think any deep investigation is needed in order to make an assertion. In fact, without making an _explicit_ assertion, we are making an _implicit_ one. But, the trace will not be friendly in case of a crash. The point here is that `mNoteList` being an `ArrayList` can have any element as `null`, even when it is not empty. In Java, when considering an `ArrayList<T>`, we not only must consider the arrayList itself as `@Nullable`, but the collection items as well (i.e. `T` is `@Nullable`). Rather than chasing down every possible insertion / mutation of this list so prove to ourselves it might be `@NonNull` (which could be invalidated by later code changes anyway), the suggestion I'm making is that we do _not_ try to go that deep.

Given that we _know_ that `mNoteList.get(position)` might be `null`, and that `getDetailFragmentForNote` accepts only `@NonNull` arguments, it is proper to do this assertion _before_ invoking the annotated method. Otherwise, we get this confusing situation where the trace points us to [this line](https://github.com/wordpress-mobile/WordPress-Android/commit/69601155de97c4bcce6fbe4dbba18cb3beea73d3#diff-e48edd5a3728c682d33322513d5ae0d968c321d8cda405d1dd45e9f4ac8b4152R404) for a `NullPointerException`, the `@NonNull` parameter annotation is clearly violated, and whoever is analyzing this crash has to work backwards and re-establish the actual source of the unexpected nullness, though we have already worked that out. I'd rather _fail-fast_ here, conveying our knowledge to the would-be crash analyst.

Reply:
👋 @mkevins and let me preface this response of mine with the fact that I don't feel strongly about adding `Objects.requireNonNull(mNoteList.get(position)`, nor it is hard for me to go and add this commit, it will take me just a minute, I just want to understand your points to the fullest so I can be more efficient going forward. 🙏 

-----

> I don't agree that adding this assertion means that it is "becoming a rule", and that doing so requires adding a similar check everywhere that something appears similar throughout the codebase.

First of all thanks for disagreeing with me Matt , this means you care and that you want the best possible outcome from this PR, and in general that you care for the overall success of the project, much appreciated! 🥇 

> This kind of rationale would prevent us from making small changes toward a larger goal.

I understand that and can agree with you on it. 💯

>  I think we should focus on this instance, and avoid prematurely generalizing.

Having said the about, I want to share with you my rational and how I ended-up with this generalizing, not being sure if that is premature or not, let me know if you still believe so, I would really appreciate your feedback here:

1. I first searched for `override fun getItem\(position\: Int\)\: Fragment \{` and this got me the below list of usages:
    1. [CommentDetailFragmentAdapter.java](https://github.com/wordpress-mobile/WordPress-Android/blob/44a6d0611175f132ff6920b4d6a3e16d565259b6/WordPress/src/main/java/org/wordpress/android/ui/comments/CommentDetailFragmentAdapter.java#L36-L40) -> `getComment(position)` -> if check on `position` -> `mCommentList.get(position)` 🤔
    2. [HistoryDetailContainerFragment.java](https://github.com/wordpress-mobile/WordPress-Android/blob/44a6d0611175f132ff6920b4d6a3e16d565259b6/WordPress/src/main/java/org/wordpress/android/ui/history/HistoryDetailContainerFragment.java#L356-L359) -> `mRevisions.get(position)` 🎯
    3. [MediaPreviewActivity.java](https://github.com/wordpress-mobile/WordPress-Android/blob/44a6d0611175f132ff6920b4d6a3e16d565259b6/WordPress/src/main/java/org/wordpress/android/ui/media/MediaPreviewActivity.java#L395-L423) -> `mMediaIdOrUrlList.get(position)` 🎯
    4. [EditPostActivity.java](https://github.com/wordpress-mobile/WordPress-Android/blob/44a6d0611175f132ff6920b4d6a3e16d565259b6/WordPress/src/main/java/org/wordpress/android/ui/posts/EditPostActivity.java#L2361-L2412) -> switch case on `position` 🤔
    5. [ReaderPhotoViewerActivity.java](https://github.com/wordpress-mobile/WordPress-Android/blob/44a6d0611175f132ff6920b4d6a3e16d565259b6/WordPress/src/main/java/org/wordpress/android/ui/reader/ReaderPhotoViewerActivity.java#L283-L288) -> `mImageList.get(position)` 🎯 
    6. [ReaderPostPagerActivity.java](https://github.com/wordpress-mobile/WordPress-Android/blob/44a6d0611175f132ff6920b4d6a3e16d565259b6/WordPress/src/main/java/org/wordpress/android/ui/reader/ReaderPostPagerActivity.java#L1012-L1027) -> if check on `position` -> `mIdList.get(position)` 🤔
    7. [ReaderSubsActivity.java](https://github.com/wordpress-mobile/WordPress-Android/blob/44a6d0611175f132ff6920b4d6a3e16d565259b6/WordPress/src/main/java/org/wordpress/android/ui/reader/ReaderSubsActivity.java#L548-L551) -> `mFragments.get(position)` 🎯 
2. I then searched for: `override fun getItem\(position\: Int\)\: Fragment \{`
    1. [PagesFragment.kt](https://github.com/wordpress-mobile/WordPress-Android/blob/44a6d0611175f132ff6920b4d6a3e16d565259b6/WordPress/src/main/java/org/wordpress/android/ui/pages/PagesFragment.kt#L663-L667) -> `pageTypes[position]` 🎯
    2. [SubfilterPageFragment.kt](https://github.com/wordpress-mobile/WordPress-Android/blob/44a6d0611175f132ff6920b4d6a3e16d565259b6/WordPress/src/main/java/org/wordpress/android/ui/reader/subfilter/SubfilterPageFragment.kt#L158-L162) -> `filterCategory[position]` 🎯

🎯 -> I think it is safe to assume that this fits our case, thus can be consider generalizing.
🤔 -> Maybe, doesn't fit to our case, thus can be considered an exception to the rule, but I am not 100% sure without making a deeper dive.

> I don't think any deep investigation is needed in order to make an assertion. In fact, without making an explicit assertion, we are making an implicit one. But, the trace will not be friendly in case of a crash. 

I agree with that and the trace being not friendly. Yet, do you think we should strive to get a friendly NPE trace in every such case? I am just trying to find the pattern here so I can apply it to my future PRs.

> The point here is that mNoteList being an ArrayList can have any element as null, even when it is not empty. In Java, when considering an ArrayList<T>, we not only must consider the arrayList itself as @Nullable, but the collection items as well (i.e. T is @Nullable).

I understand the technicalities here for a general Java based `ArrayList<T>`, but do you think this applies to `FragmentStatePagerAdapter`, `FragmentPagerAdapter`, `PagerAdapter` and its API contract? 🤔

> Rather than chasing down every possible insertion / mutation of this list so prove to ourselves it might be @NonNull (which could be invalidated by later code changes anyway), the suggestion I'm making is that we do not try to go that deep.

I agree here and I would suggest the same.

> Given that we know that mNoteList.get(position) might be null, and that getDetailFragmentForNote accepts only @NonNull arguments, it is proper to do this assertion before invoking the annotated method.

Ah, please help me understand this bit, how do we know that `mNoteList.get(position)` might be null? Do you mean we know that because of Java and the fact that `ArrayList<T>` can accept nullable items, or because you indeed have identified cases when the `FragmentStatePagerAdapter`, `FragmentPagerAdapter`, `PagerAdapter` and its API can produce such case, possibly valid cases? 🤔

> Otherwise, we get this confusing situation where the trace points us to [this line](https://github.com/wordpress-mobile/WordPress-Android/commit/69601155de97c4bcce6fbe4dbba18cb3beea73d3#diff-e48edd5a3728c682d33322513d5ae0d968c321d8cda405d1dd45e9f4ac8b4152R404) for a NullPointerException, the @NonNull parameter annotation is clearly violated, and whoever is analyzing this crash has to work backwards and re-establish the actual source of the unexpected nullness, though we have already worked that out.

I agree this is confusing.

> I'd rather fail-fast here, conveying our knowledge to the would-be crash analyst.

👍

Reply:
> First of all thanks for disagreeing with me Matt , this means you care and that you want the best possible outcome from this PR, and in general that you care for the overall success of the project, much appreciated! 🥇

Of course, and I feel the same way toward your position on these discussions. 👍 

> I agree with that and the trace being not friendly. Yet, do you think we should strive to get a friendly NPE trace in every such case? I am just trying to find the pattern here so I can apply it to my future PRs.

> I understand the technicalities here for a general Java based ArrayList<T>, but do you think this applies to FragmentStatePagerAdapter, FragmentPagerAdapter, PagerAdapter and its API contract? 🤔

No, I think it is quite possible that there exists a water-tight argument that the item in these other cases (and even possibly in _this_ case) is guaranteed to be `@NonNull`. But, we have not annotated it as such, so the suggestion I'm making is purely a shallow suggestion to improve a trace (albeit one that may never occur).

> Ah, please help me understand this bit, how do we know that mNoteList.get(position) might be null? Do you mean we know that because of Java and the fact that ArrayList<T> can accept nullable items, or because you indeed have identified cases when the FragmentStatePagerAdapter, FragmentPagerAdapter, PagerAdapter and its API can produce such case, possibly valid cases? 🤔

The first one. Since we aren't strictly taking about framework APIs (i.e. this nullity is governed by our own code, it is _possible_ we haven't considered every case).

From the discussion, I wonder if you consider adding `Objects.requireNotNull` as some equivalent to annotating as `@Nullable`? 🤔  I don't feel strongly enough about adding it to block this PR (hence why already approved 😄 ), but this discussion is useful anyway, imo, so I'll explain how I'm thinking about it:

I see `@Nullable` and `@NonNull` annotations as trail markers, delineating the boundaries between the world of `null` and the world of no more NPEs. In the world of `null`, it's quite possible to have no NPEs, but requires constant diligence on behalf of developers, and often a defensive programming mindset that can lead to redundancy in guard code, potential dead code, and more complexity due to branching. In the world of `@NonNull`, NPEs are not even possible, and our code can be elegant, expressive, and concise.

Imo, it would be great to shrink the `@Nullable` world down to as small a surface area as possible, and expand the `@NonNull` area. We are like park rangers, maintaining the trail markers, and when we want to convert an area to `@NonNull`, we typically have to choose between gracefully guarding, or checking and crashing. To me, `requireNotNull` acts as a gateway to allow us to connect the unmarked trails that we encounter with our growing network of `@NonNull` trails. If we are going to crash, we know exactly where.

All that said, I think I understand where you are coming from in that this is like a "model" PR, and I understand your hesitation to include an "ad hoc" check.